### PR TITLE
As3935 calibration

### PR DESCRIPTION
**Pull request by mveinot**
# What does this implement/fix?

Pick-up feature started in https://github.com/esphome/esphome/pull/2084/files (implement antenna calibration and tuning for the AS3935 lightning detector) and attempt to bring them inline with requested changes

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

https://github.com/esphome/esphome/pull/2084

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
as3935_spi:
  cs_pin: GPIO5
  irq_pin:
    number: GPIO15
    mode: INPUT_PULLDOWN
  indoor: False
  noise_level: 2
  spike_rejection: 5      # Default 2
  lightning_threshold: 1
  mask_disturber: False
  div_ratio: 0
  capacitance: 2
  watchdog_threshold: 2
  tune_antenna: False
  calibration: True
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/as3935/as3935.cpp [hunk 1]
```diff
@@ -227,6 +233,87 @@ uint32_t AS3935Component::get_lightning_energy_() {
...
+  }
+}
+
+void AS3935Component::tune_antenna() {
+  ESP_LOGI(TAG, "Starting antenna tuning...");
+  uint8_t div_ratio = this->read_div_ratio();
+  uint8_t tune_val = this->read_capacitance();
+  ESP_LOGI(TAG, "Division Ratio is set to: %d", div_ratio);
+  ESP_LOGI(TAG, "Internal Capacitor is set to: %d", tune_val);
+  ESP_LOGI(TAG, "Displaying oscillator on INT pin. Measure its frequency - multiply value by DivisionRatio");
```
Comment (drawgas) [inline esphome/components/as3935/as3935.cpp]: Neet picking, but `DivisionRatio` => `Division Ratio` just to be consistent

Comment (mveinot) [reply esphome/components/as3935/as3935.cpp]: Fix applies

**Review by drawgas — APPROVED** [review]
Looks good to me

File: esphome/components/as3935/as3935.cpp [hunk 2]
```diff
@@ -32,6 +32,12 @@ void AS3935Component::dump_config() {
 #ifdef USE_SENSOR
   LOG_SENSOR("  ", "Distance", this->distance_sensor_);
   LOG_SENSOR("  ", "Lightning energy", this->energy_sensor_);
+  if (this->tune_antenna_) {
+    ESP_LOGCONFIG(TAG, "  Antenna tuning: ENABLED - lightning detection will not function in this mode");
+    this->tune_antenna();
+  } else if (this->calibration_) {
+    this->calibrate_oscillator();
+  }
```
Comment (jesserockz) [inline esphome/components/as3935/as3935.cpp]: This "work" should not be done in `dump_config`, as per the name, its for dumping config, not for performing functions. This should be moved to `setup`

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/as3935/as3935.cpp (hunk 2 above)
Comment (mveinot) [reply esphome/components/as3935/as3935.cpp]: Makes perfect sense, moved it.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**